src/a.tex

```diff
@@ -157,29 +159,49 @@ \section{Load-Reserved/Store-Conditional Instructions}
-and the SC, and if there is no other SC between the LR and itself in program
-order.
-Note this LR might have had a different address argument and data size, but
-reserved the SC's address as part of the reservation set.
-Following this model, in systems with memory translation, an SC is allowed to
-succeed if the earlier LR reserved the same location using an alias with
-a different virtual address, but is also allowed to fail if the virtual
-address is different.
-The SC must fail if the address is not within the reservation set of the most
-recent LR in program order.
-The SC must fail if a store from another hart, or a write from some other
-device, to the reservation set can be observed to occur between the LR and the
-SC.
-An SC must fail if there is another SC (to any address) between the LR and the
-SC in program order.
-The precise statement of the atomicity requirements for successful LR/SC
-sequences is defined by the Atomicity Axiom in Section~\ref{sec:rvwmo}.
+The reservation set accessed by the LR and SC instructions must be {\em
+reservable}, i.e., the addresses in the reservation set must support these
+operations; otherwise, executing an LR or SC instruction generates an access
+exception.
+
+An LR registers an arbitrarily large reservation set, and the reservation set
+must include all bytes of the addressed data word or doubleword.  An SC
+succeeds if all of the following conditions are true when the SC is executed:
+\vspace{-0.2in}
+\begin{itemize}
+\parskip 0pt
+\itemsep 1pt
+\item The reservation is valid.
+\item The reservation set includes all of the bytes of the word or doubleword
+addressed by the SC.
+\end{itemize}
+If any of the above conditions are false when the SC is executed, the SC fails.
+
+The following events are guaranteed to invalidate the reservation on hart
+{\em H}:
+\vspace{-0.2in}
+\begin{itemize}
+\parskip 0pt
+\itemsep 1pt
+\item {\em H} executes a successful SC.
+\item {\em H} executes an SC that addresses bytes that are not included in the
+reservation set.
+\item A memory write (e.g., an unconditional store, an AMO, a
+successful SC, or a device write) performed by another hart or device to a byte
+included in the reservation set is observed between the LR and the SC executed
+by {\em H}.
+\end{itemize}
+
+Architectural specifications that extend the base specification, e.g.
+extensions, privileged architecture, and platform specifications, may identify
+additional conditions that cause the reservation to be invalidated.
+
+In general, an SC in an LR/SC sequence may succeed only if certain atomicity
+requirements, as defined by the Atomicity Axiom in Section 14.1, are satisfied.
+If these requirements cannot be satisfied, the SC in an LR/SC sequence must
+fail. In particular, if two harts do not agree on the memory order of accesses
+to a reservation set, the atomicity of accesses from those two harts to that
+reservation set cannot be maintained, and so the SC must fail.
```

Comment:
@daniellustig your thoughts on this paragraph?

Reply:
This text seems to suggest that after checking if the reservation is still valid, there is some kind of additional step to see if the Atomicity Axiom has also been satisfied.  I'd say its largely the opposite: the Atomicity Axiom just formalizes the atomicity semantics that the reservation already enforces for successful SCs.

I'd also quibble about the wording about harts agreeing on memory order, but before we go there...what would you all say to just deleting this paragraph altogether?

Reply:
There are a few things we were trying to do here.  The first was to expand on the existing reference to the Atomicity Axiom:

"The precise statement of the atomicity requirements for successful LR/SC sequences is defined by the Atomicity Axiom in Section 14.1."

The second was to point out that the SC can only succeed if the Atomicity Axiom holds; otherwise, the SC must fail. Upon reflection, I see this is a clumsy attempt to tie this back to the memory model, especially since a failed SC won't appear in the global memory order, rendering the axiom moot. So yeah, I'd say take it out.

The third was to introduce a notion of "big-C Coherence" that was floating around in the discussion groups. Again, I was attempting to tie this back to the memory model, and I was trying to create an architectural "litmus test" for events that must cause an SC to fail that could apply once more than the base spec was considered. For example, differences in or modifications to the PTEs and PMAs could break big-C Coherence between harts, because each hart accesses the reservation set in a different manner, so there should be no reasonable expectation of success. (Spoiler alert: I'm anticipating the necessity to describe the behavior of accesses with mixed attributes at some point in the future, so I'm trying to lay the groundwork now....)

Independent of this particular text, I do think there needs to be a formal definition of the coherence order, which is related to the global memory order, and which should be a function of PTEs/PMAs/etc. once the various extended architecture specs are considered, and it seems clear that if two harts can't observe the same coherence order for memory locations in the reservation set, they can't use LR/SC between them. Maybe that's a topic for a different day, but I'd like not to kick the can down the road too much longer....

Reply:
I don't think it's helpful to sneak in a new concept of coherence just anticipating that it might later be used somehow.  Can you give an example of where you think the existing global memory order formulation is incomplete?

Reply:
Fair point. Perhaps the overall discussion should be reserved (so to speak....) for another time and forum.

To answer your question, I think there's an assumption that all harts have the same architectural state with respect to how they access memory (e.g. memory attributes) and that assumption ensures that the global memory order is observed by all harts. There are two (or more) things that undermine that assumption: mismatched architectural state (e.g. programming memory attributes differently) and changing architectural state (e.g. modifying memory attributes at runtime, which is kinda mismatched architectural state, but in a dynamic sense). In my experience, the abstraction that all agents access memory in the same big-C Coherent manner is not practical (e.g. due to different capabilities of agents) or even desirable (e.g. enforcing cache coherence may negatively impact system cost or performance). In addition, changing the architectural state that affects how memory is accessed is never atomic across all agents, so during those transition periods, different agents can in principle observe different memory orders. (One can carefully craft software sequences to avoid this; I'm aware of these sequences in other architectures, but I'm uncertain whether they exist in RV yet.)

(Note that I don't believe I'm the first to raise these issues, and my understanding is that some of these would be covered in MM2.0, no?)

For the purposes of LR/SC, I was trying to express the notion that LR/SC can't be atomic if harts don't observe the same memory order, at a minimum, on the subset of the global memory order pertaining to the addresses in the reservation set. (LR/SC may still be atomic if harts observed different memory orders on addresses not in the reservation set, however.) But perhaps introducing the tip of the iceberg in the LR/SC section is the wrong move.

Long story short, I think the memory model description and the behavior of LR/SC is consistent in the absence these issues. So for now, I'm OK removing the text Andrew highlighted in his comment here.

Reply:
OK... I'm going to refrain from the long speech I would have put in here, other than to say that I think forward referencing to 14 here is a bad idea (so yes, Dan cringes every time I say it -- I'm with him here).

For the record, I believe the axiomatic model in 14 does include the notion of Big-C coherence. You take the global memory order and then restrict it to writes to the same location. Out pops what other axiomatic memory models call the "co" or coherence order which is the total order over all writes to a given location and is equivalent to the notion I called "big-C" coherence. 

But I think the rules for LR/SC succeeding operationally should ALL be clearly in the A extension text without needing 14 (i.e. we don't want an implementer to have to swallow and understand, or really even read, 14's much more abstract mathematical worldview in order to build the right hardware for LR/SC).

Now, I hear you about the PMAs moving about and in particular the coherence PMA mismatching. And this is most definitely a MM2.0 discussion, but at the point you drop coherence, I don't think you can reasonably define an architecture to LR/SC. You can define a lot of different implementations, but I don't see how you make a one-size fits all architecture for LR/SC work.

Derek

Reply:
@strikerdw I agree with you, but I have two additional comments. Number 1: It might be useful to explicitly define coherence order as you described it (I agree with that definition). Number 2: There has to be a consistent memory order on the entire reservation set which is more expansive, I think, than the coherence order on the individual memory locations within the set.

Maybe 2 is not worth the virtual ink because, at the end of the day, the eventual success guarantee requires the LR and SC accesses to be to the same memory location (Coherence!), but I think there is a subtle difference in the "pure" architectural definition of LR/SC.